depthai_ros_driver/src/depthai_base_ros2.cpp

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <depthai_ros_driver/depthai_base_ros2.hpp>
+
+namespace rr {
+
+//==============================================================================
+DepthAIBaseRos2::DepthAIBaseRos2(const rclcpp::NodeOptions& options)
+: Node("depthai_node", options)
+{
+  auto get_param = [this](const std::string& name, auto& variable)
+    {
+      using var_type = std::remove_reference_t<decltype(variable)>;
+      if (this->has_parameter(name))
+        this->get_parameter(name, variable);
+      else
+        this->declare_parameter<var_type>(name, variable);
+    };
+
+  get_param("queue_size", _queue_size);
+  get_param("camera_name", _camera_name);
+  get_param("camera_param_uri", _camera_param_uri);
+
+  if (_camera_param_uri.back() != '/')
+    _camera_param_uri += "/";
+
+  _depthai_common = std::make_unique<DepthAICommon>(get_param);
+
+  _depthai_common->register_callbacks(
+    [this](const HostDataPacket& packet, Stream type, double ts)
+    {
+      this->publishImageMsg(packet, type, this->get_rostime(ts));
+    },
+    [this](const dai::Detections& detections, double ts)
+    {
+      this->publishObjectInfoMsg(detections, this->get_rostime(ts));
+    }
+  );
+
+  _topic_names = _depthai_common->get_topic_names();
+
+  /// ros2 dds qos profile
+  const auto driver_qos = rclcpp::ServicesQoS().best_effort();
+
+  // create service
+  _camera_info_default = this->create_service<TriggerSrv>(
+    "reset_camera_info",
+    [this](
+      const std::shared_ptr<TriggerSrv::Request> req,
+      std::shared_ptr<TriggerSrv::Response> res)
```

Comment:
Does ROS2 limit services to be shared_ptr based (similar to how ROS1 restricts them to be references)?